File: 0074-search-a-2d-matrix-搜索二维矩阵/0074.py
# 二分查找
# 将矩阵转化为一维序列

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        beg = 0
        end = len(matrix) * len(matrix[0])
        while beg < end:
            mid = (beg + end) // 2
            midY = mid // len(matrix[0])
            midX = mid % len(matrix[0])
            if matrix[midY][midX] == target:
                return True
            if matrix[midY][midX] > target:
                end = mid
            else:
                beg = mid + 1
        
        return False

File: 0074-search-a-2d-matrix-搜索二维矩阵/0074.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.size() == 0) return false;
        int h = matrix.size();
        int w = matrix[0].size();
        if (w == 0) return false;
        
        int beg = 0;
        int end = h * w - 1;
        int mid;
        while(beg < end) {
            mid = (end + beg) / 2;
            int y = mid / w;
            int x = mid % w;
            if (matrix[y][x] >= target) {
                end = mid;
            } else {
                beg = mid + 1;
            }
        }
        return matrix[beg / w][beg % w] == target;
    }
};